### ISSUE TITLE
parsing of concatenation with initial optional sequence fails incorrectly

### ISSUE THREAD

**Issue by matthewleon** [ldthomas/apg-js#7]
This is a follow up to https://github.com/ldthomas/apg-js/issues/6

Take this example grammar:
```
const GRAMMAR = String.raw`full = [ab] b
ab = "a" / "b"
b = "b"
`;
```

and test string:
```
const EXAMPLE = "b";
```

The string should parse successfully. It does not, due to the same reasoning as the issue above. `[ab]` produces a match, and the parser never seems to "back up" and consider whether rule `b` alone can produce a success, without the application of `[ab]`.

**Comment by ldthomas** [comment]
You are correct. The parser never backtracks to look for other phrases. That is a feature, not a bug. It does "greedy" repetitions and "prioritized choice" selection of alternates like PEG parsers. APG has worked like this since even before I knew of PEG because it 1) simplifies the problem of selecting a single phrase from the "forest" of parse trees and 2) once you get used to it, for me at least, it greatly simplifies the problem of creating a grammar to match the phrases your problem demands and 3) backtracking is inefficient and is the source of the "catastrophic backtracking" problem of regex. I've modified a number of RFC grammars to work with these rules and I've rarely run into a situation that couldn't be rewritten to work with this single choice feature. For example, your problem 

full = ["a" / "b"] "b"

only matches three possible phrases, "ab", "bb" and "b". So simply rewrite the grammar as

full = "bb" / "ab" / "b"

and you are there. The only rule you really need is to remember to always put the longest phrase first (left-most).

In more complex cases, I've occasionally been forced to use look ahead operators, which I did steal from Parr and Ford (https://en.wikipedia.org/wiki/Syntactic_predicate) (but I've always tried to be careful and give them credit for them) but they are every bit as inefficient as backtracking (in fact they always backtrack) so I also avoid them when at all possible. Your example above is a simple illustration of the point, but for a more complex example have a look at my ABNF grammar for XML(https://github.com/ldthomas/apg-7.0/blob/main/xml/xml.abnf). The XML specification (https://www.w3.org/TR/xml/) has many productions which cry out for lookahead operators and my first go at it used quite a few. But on closer examination I found that even there I could eliminate them all. Some of the phrase matching is handled semantically in callback functions, but the grammar itself is pure ABNF with no lookahead or other special operators.

Maybe it's just me, but give it a try, writing your grammars with prioritized choice and greedy repetitions in mind and see if you don't warm up to it.
